providers/azure.go

```diff
@@ -230,44 +276,48 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 		err = otherMailsErr
 	}
 
+	if err != nil || email == "" {
+		email, err = json.Get("userPrincipalName").String()
+		if err != nil {
+			logger.Errorf("unable to find userPrincipalName: %s", err)
+			return "", err
+		}
+	}
+
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	var email string
-	var err error
-
-	if s.AccessToken == "" {
+func getEmailFromProfileAPI(ctx context.Context, accessToken, profileURL string) (string, error) {
+	if accessToken == "" {
 		return "", errors.New("missing access token")
 	}
 
-	json, err := requests.New(p.ProfileURL.String()).
+	json, err := requests.New(profileURL).
 		WithContext(ctx).
-		WithHeaders(makeAzureHeader(s.AccessToken)).
+		WithHeaders(makeAzureHeader(accessToken)).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		return "", err
 	}
 
-	email, err = getEmailFromJSON(json)
-	if err == nil && email != "" {
-		return email, err
-	}
+	return getEmailFromJSON(json)
+}
 
-	email, err = json.Get("userPrincipalName").String()
-	if err != nil {
-		logger.Errorf("failed making request %s", err)
-		return "", err
+// EnrichSessionState finds the email to enrich the session state
+func (p *AzureProvider) EnrichSessionState(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
 	}
 
+	email, err := getEmailFromProfileAPI(ctx, s.AccessToken, p.ProfileURL.String())
 	if email == "" {
-		logger.Errorf("failed to get email address")
-		return "", err
+		logger.Errorf("failed to get email address: %s", err)
```

Comment:
This feels like a bug (that looks like it might've been there before): the `email == ""` check is after the `err != nil` check, but in the `email == ""` check we return `err` (which is `nil`).

I think we want to make our own custom `errors.New("...")` here if the email is blank right? Do we want to explicitly call the `logger.Errorf` here, or make a proper error with the details and let the logging get handled when the error is logged?

Reply:
yeah, this logic looks weird, I feel it should be 
```go
if email == "" || err != nil {
```

Reply:
returning new error will be a change to existing behavior. I think I'm fine with it.